hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DFSClient.java

```diff
@@ -3019,4 +3013,52 @@ private void checkSmallFilesSupportConf(Configuration conf) throws IOException {
   public boolean hasLeader(){
     return leaderNN!=null;
   }
+  
+  public void addUserGroup(final String userName, final String groupName)
+      throws IOException {
+    try{
+      namenode.addUserGroup(userName, groupName, false);
+    }catch (RemoteException re){
+      throw re.unwrapRemoteException();
+    }
+    
+    if(userName != null && groupName != null){
```

Comment:
Why should we multicast to all NNs for adding? If the entry is not in cache, CacheLoader will fetch it from DB, right? I suppose it's a matter of what is more performant, RPC or go to DB.

Reply:
It is true only for the addition of a user or a group, but not for adding a user to a group. 
For example, if we have two namenodes, and both have the following mapping  "mahmoud" -> "hdfs", and then from the client we added mahmoud to another group "yarn",  then only the NN serving the add request will be able to update its cache to "mahmoud" -> "hdfs, yarn", however, for the other namenode it has to wait until the cache entry is evicted to know about the new changes.